File: packages/plugin-sdk/assembly/crypto.ts
import {
  peekoo_crypto_ed25519_get_or_create,
  peekoo_crypto_ed25519_sign,
} from "./host";
import { extractStringField, quote } from "./json";
import { readString, writeString } from "./memory";

export class Ed25519PublicKeyInfo {
  constructor(
    public publicKeyBase64Url: string,
    public publicKeySha256Hex: string,
  ) {}
}

export function ed25519GetOrCreate(alias: string): Ed25519PublicKeyInfo {
  const input = "{\"alias\":" + quote(alias) + "}";
  const offset = peekoo_crypto_ed25519_get_or_create(writeString(input));
  const response = readString(offset);
  return new Ed25519PublicKeyInfo(
    extractStringField(response, "publicKeyBase64Url"),
    extractStringField(response, "publicKeySha256Hex"),
  );
}

export function ed25519Sign(alias: string, payload: string): string {
  const input = "{\"alias\":" + quote(alias) + ",\"payload\":" + quote(payload) + "}";
  const offset = peekoo_crypto_ed25519_sign(writeString(input));
  return extractStringField(readString(offset), "signatureBase64Url");
}

File: crates/peekoo-plugin-sdk/src/crypto.rs
//! Cryptography helpers for plugins.
//!
//! Requires the `crypto:ed25519` permission.
//! Keys are persisted per plugin and namespaced by alias.

use extism_pdk::{Error, Json};

use crate::host_fns::{
    peekoo_crypto_ed25519_get_or_create, peekoo_crypto_ed25519_sign,
    CryptoEd25519GetOrCreateRequest, CryptoEd25519SignRequest,
};

#[derive(Clone, Debug)]
pub struct Ed25519PublicKeyInfo {
    /// Public key encoded as URL-safe base64 without padding.
    pub public_key_base64_url: String,
    /// SHA-256 digest of the public key in lowercase hex.
    pub public_key_sha256_hex: String,
}

/// Loads or creates a plugin-scoped Ed25519 keypair for the provided alias.
pub fn ed25519_get_or_create(alias: &str) -> Result<Ed25519PublicKeyInfo, Error> {
    let response = unsafe {
        peekoo_crypto_ed25519_get_or_create(Json(CryptoEd25519GetOrCreateRequest {
            alias: alias.to_string(),
        }))?
    };

    Ok(Ed25519PublicKeyInfo {
        public_key_base64_url: response.0.public_key_base64_url,
        public_key_sha256_hex: response.0.public_key_sha256_hex,
    })
}

/// Signs a UTF-8 payload with the plugin-scoped Ed25519 key for `alias`.
pub fn ed25519_sign(alias: &str, payload: &str) -> Result<String, Error> {
    let response = unsafe {
        peekoo_crypto_ed25519_sign(Json(CryptoEd25519SignRequest {
            alias: alias.to_string(),
            payload: payload.to_string(),
        }))?
    };

    Ok(response.0.signature_base64_url)
}
